src/Core/Domain/Product/ValueObject/ProductRedirectOption.php

```diff
@@ -0,0 +1,227 @@
+     *
+     * @return static
+     *
+     * @throws LogicException
+     */
+    public static function buildRedirect(string $redirectType, int $redirectTargetId): self
+    {
+        if ($redirectType === self::TYPE_NO_REDIRECT) {
+            throw new LogicException(sprintf('Use "%s"::buildNoRedirect for building "no_redirect" option', self::class));
+        }
+
+        return new self($redirectType, $redirectTargetId);
+    }
+
+    /**
+     * @return string
+     */
+    public function getRedirectType(): string
+    {
+        return $this->redirectType;
+    }
+
+    /**
+     * @return int
+     */
+    public function getRedirectTargetId(): int
+    {
+        return $this->redirectTargetId;
+    }
+
+    /**
+     * Use static factories to initiate this class
+     *
+     * @param string $redirectType
+     * @param int $redirectTargetId
+     *
+     * @throws ProductConstraintException
+     */
+    private function __construct(string $redirectType, int $redirectTargetId)
+    {
+        $this->assertRedirectType($redirectType);
+        $this->assertTypeAndIdIntegrity($redirectType, $redirectTargetId);
+        $this->redirectType = $redirectType;
+        $this->redirectTargetId = $redirectTargetId;
+    }
+
+    /**
+     * @param string $value
+     *
+     * @throws ProductConstraintException
+     */
+    private function assertRedirectType(string $value): void
+    {
+        if (in_array($value, self::AVAILABLE_REDIRECT_TYPES)) {
+            return;
+        }
+
+        throw new ProductConstraintException(
+            sprintf(
+                'Invalid product redirect type "%s". Allowed types are: "%s"',
+                $value,
+                implode(',', self::AVAILABLE_REDIRECT_TYPES)
+            ),
+            ProductConstraintException::INVALID_REDIRECT_TYPE
+        );
+    }
+
+    /**
+     * @param string $type
+     * @param int $id
+     *
+     * @throws ProductConstraintException
+     */
+    private function assertTypeAndIdIntegrity(string $type, int $id): void
+    {
+        $isProductType = $type === self::TYPE_PRODUCT_TEMPORARY || $type === self::TYPE_PRODUCT_PERMANENT;
+
+        if ($isProductType) {
+            $this->assertProductId($id);
+
+            return;
+        }
+
+        $this->assertCategoryId($id);
+    }
+
+    /**
+     * @param int $id
+     *
+     * @throws ProductConstraintException
+     */
+    private function assertProductId(int $id): void
+    {
+        if ($id <= 0) {
+            throw new ProductConstraintException(
+                sprintf('Invalid product redirect target id "%s". It is required when redirecting to product', $id),
+                ProductConstraintException::INVALID_REDIRECT_TARGET_ID
+            );
+        }
+    }
+
+    /**
+     * @param int $id
+     *
+     * @throws ProductConstraintException
+     */
+    private function assertCategoryId(int $id): void
+    {
+        if ($id !== self::NO_TARGET_VALUE && $id <= 0) {
```

Comment:
You shouldn't check `self::NO_TARGET_VALUE` when category redirection is set, it's only used for `TYPE_NO_REDIRECT` so this special use case feels weird it's equivalent to `$id !== 0 && $id <= 0` which is like `$id < 0`
It would be more logical remove this check and call `assertCategoryId` only when redirection type is `TYPE_CATEGORY_PERMANENT` or `TYPE_CATEGORY_TEMPORARY`

Reply:
Unless `NO_TARGET_VALUE` is used to redirect to parent category? then maybe it's worth adding a dedicated const (event if it also contains 0) so that it's self explanatory

Reply:
The latest concept from https://github.com/PrestaShop/PrestaShop/pull/20554 should solve this. @jolelievre if I am not wrong here is the right place to introduce a `NoCategoryId` class ?

Reply:
yup its the case `If no category is selected, the Main Category is used`.

Reply:
So we have these cases:
1. No redirect -> we create new VO ProductNoRedirectOption which extends ProductRedirectOption and avoids parent constraints with forced values (as in #20554).
2. Category redirect (when category id is 0) -> whats here? (it is still a RedirectOption, just indicating that it uses some default category for redirection) - can we introduce something like "ProductCategoryDefaultRedirectOption"? Isn't it overengineering? its almost if we soon gonna have 1class=1option. 
3. all other types -> ProductRedirectOption($type, $targetId);

Then in SetRedirectOption (when building command) decide which object should be initiated. And all of that just to ensure command validity before reaching handler?

Reply:
Its hard to come by a good solution. If we create 3 separate objects(e.g NoRedirectOption, RedirectToCategoryOption, RedirectToProductOption) we cannot typehint it in command, which seems even worse than having previous implementation (Having abstract class for typehint or an interface for VO also doesn't sound good). So I wait for @eternoendless decision on this.

Reply:
We've decided to use simpliest solution with @matks. I wont use the VO at all, pass the redirectType and targetId to command and simply validate it inside the handler.

Reply:
so finally, I've came up to a solution which you could see in this commit https://github.com/PrestaShop/PrestaShop/pull/20473/commits/de28e0b8c6f0496517a718023809f0fca2087698.
TLDR: added RedirectType and RedirectTarget VO's and wrapped them in RedirectOption which ensures they integrity before reaching handler. By far it seems the best solution for me